pkg/test/environment.go

```diff
@@ -80,6 +80,12 @@ func NewEnvironment(scheme *runtime.Scheme, options ...functional.Option[Environ
 		// https://github.com/kubernetes/enhancements/issues/2249 for more info.
 		environment.ControlPlane.GetAPIServer().Configure().Set("feature-gates", "PodAffinityNamespaceSelector=true")
 	}
+	// MinDomainsInPodTopologySpread enforces a minimum number of eligible node domains for pod scheduling
+	// See https://kubernetes.io/docs/concepts/scheduling-eviction/topology-spread-constraints/#spread-constraint-definition
+	// Ref: https://github.com/aws/karpenter-core/pull/330
+	if version.Minor() >= 24 {
```

Comment:
It can only be set starting in 1.24, before that it would error out as the flag doesn't exist.  I *think* we should only set it for 1.27+ since that is where it is defaulted to true upstream.  It's good that we test it both ways just to ensure that having the flag off doesn't somehow break some existing logic.

Reply:
Doh, just realized MinDomainsInPodTopologySpread is a kubelet feature and not an apiserver feature. Looking through controller-runtime right now to see if its envtest package supports this, haven't found it yet...

Reply:
It should be both I think.

Reply:
apiserver panics when starting up w/ MinDomainsInPodTopologySpread feature-gate config if you manually enforce v1.24, for example:

```
$ git diff
diff --git a/pkg/test/environment.go b/pkg/test/environment.go
index f5a32cd..5385c72 100644
--- a/pkg/test/environment.go
+++ b/pkg/test/environment.go
@@ -18,7 +18,6 @@ import (
        "context"
        "log"
        "os"
-       "strings"
 
        "github.com/samber/lo"
        corev1 "k8s.io/api/core/v1"
@@ -31,7 +30,6 @@ import (
        "sigs.k8s.io/controller-runtime/pkg/client"
        "sigs.k8s.io/controller-runtime/pkg/envtest"
 
-       "github.com/aws/karpenter-core/pkg/utils/env"
        "github.com/aws/karpenter-core/pkg/utils/functional"
 )
 
@@ -71,7 +69,7 @@ func NewEnvironment(scheme *runtime.Scheme, options ...functional.Option[Environ
        ctx, cancel := context.WithCancel(context.Background())
 
        os.Setenv(system.NamespaceEnvKey, "default")
-       version := version.MustParseSemantic(strings.Replace(env.WithDefaultString("K8S_VERSION", "1.21.x"), ".x", ".0", -1))
+       version := version.MustParseSemantic("1.24.0")
        environment := envtest.Environment{Scheme: scheme, CRDs: opts.crds}
        if version.Minor() >= 21 {
                // PodAffinityNamespaceSelector is used for label selectors in pod affinities.  If the feature-gate is turned off,
```

yields

```
$ pwd
/Users/jackfrancis/go/src/github.com/aws/karpenter-core/pkg/controllers/provisioning/scheduling
$ go test -timeout 10m -tags e2e
Running Suite: Controllers/Scheduling - /Users/jackfrancis/go/src/github.com/aws/karpenter-core/pkg/controllers/provisioning/scheduling
=======================================================================================================================================
Random Seed: 1684188459

Will run 235 of 235 specs
------------------------------
[BeforeSuite] [PANICKED] [5.354 seconds]
[BeforeSuite] 
/Users/jackfrancis/go/src/github.com/aws/karpenter-core/pkg/controllers/provisioning/scheduling/suite_test.go:79

  [PANICKED] Test Panicked
  In [BeforeSuite] at: /Users/jackfrancis/go/pkg/mod/github.com/samber/lo@v1.38.1/errors.go:53 @ 05/15/23 15:07:44.756

  unable to start control plane itself: failed to start the controlplane. retried 5 times: timeout waiting for process kube-apiserver to start successfully (it may have failed to start, or stopped unexpectedly before becoming ready)

  Full Stack Trace
    github.com/samber/lo.must({0x25f6bc0, 0xc000110320}, {0x0, 0x0, 0x0})
    	/Users/jackfrancis/go/pkg/mod/github.com/samber/lo@v1.38.1/errors.go:53 +0x1f4
    github.com/samber/lo.Must[...](...)
    	/Users/jackfrancis/go/pkg/mod/github.com/samber/lo@v1.38.1/errors.go:65
    github.com/aws/karpenter-core/pkg/test.NewEnvironment(0xc0009818f0, {0xc0004b96f0?, 0x0?, 0x0?})
    	/Users/jackfrancis/go/src/github.com/aws/karpenter-core/pkg/test/environment.go:88 +0x4ed
    github.com/aws/karpenter-core/pkg/controllers/provisioning/scheduling_test.glob..func2()
    	/Users/jackfrancis/go/src/github.com/aws/karpenter-core/pkg/controllers/provisioning/scheduling/suite_test.go:80 +0xaf
------------------------------
[AfterSuite] [PANICKED] [0.000 seconds]
[AfterSuite] 
/Users/jackfrancis/go/src/github.com/aws/karpenter-core/pkg/controllers/provisioning/scheduling/suite_test.go:96

  [PANICKED] Test Panicked
  In [AfterSuite] at: /usr/local/opt/go/libexec/src/runtime/panic.go:260 @ 05/15/23 15:07:44.759

  runtime error: invalid memory address or nil pointer dereference

  Full Stack Trace
    github.com/aws/karpenter-core/pkg/test.(*Environment).Stop(0x0)
    	/Users/jackfrancis/go/src/github.com/aws/karpenter-core/pkg/test/environment.go:125 +0x19
    github.com/aws/karpenter-core/pkg/controllers/provisioning/scheduling_test.glob..func3()
    	/Users/jackfrancis/go/src/github.com/aws/karpenter-core/pkg/controllers/provisioning/scheduling/suite_test.go:97 +0x25
------------------------------

Summarizing 2 Failures:
  [PANICKED!] [BeforeSuite] 
  /Users/jackfrancis/go/pkg/mod/github.com/samber/lo@v1.38.1/errors.go:53
  [PANICKED!] [AfterSuite] 
  /usr/local/opt/go/libexec/src/runtime/panic.go:260

Ran 0 of 235 Specs in 5.357 seconds
FAIL! -- A BeforeSuite node failed so all tests were skipped.
--- FAIL: TestScheduling (5.36s)
FAIL
exit status 1
FAIL	github.com/aws/karpenter-core/pkg/controllers/provisioning/scheduling	7.133s
```

Reply:
There's a flag called something like AttachControlplaneOutput on the environment you can set to make it log.  I suspect the issue is that the old flag we are setting was removed in newer versions of K8s, so now it errors out if you try to set it.

Reply:
good tip!

```
Error: invalid argument "MinDomainsInPodTopologySpread=true" for "--feature-gates" flag: unrecognized feature gate: MinDomainsInPodTopologySpread
```

I don't know if we can do a foo like this in our flow:

- https://github.com/kubernetes/kubernetes/blob/v1.24.0/test/integration/scheduler/predicates_test.go#L1225

Will try tomorrow.